### PR TITLE
feat: Macos integration

### PR DESCRIPTION
**Pull request by fi-nik**
## Description
Continued work on macos integration from 
https://github.com/react-native-device-info/react-native-device-info/pull/1057#getdevicetoken

Check List before ready for merge:
- [x] Update documentation
- [ ] GitHub Action CI similar to Windows implementation
- [x] Go through each method, determine if possible for macOS & what it should return (updating table as progress is made)

| Method                                                            | Return Type         |  iOS | Android | Windows | MacOS | Web |
| ----------------------------------------------------------------- | ------------------- | :--: | :-----: | :-----: | :-----: |  :-: |
| [getAndroidId()](#getandroidid)                                   | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getApiLevel()](#getapilevel)                                     | `Promise<number>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getApplicationName()](#getapplicationname)                       | `string`            |  ✅  |   ✅    |   ✅    |  ✅  | ❌  |
| [getAvailableLocationProviders()](#getAvailableLocationProviders) | `Promise<Object>`   |  ✅  |   ✅    |   ❌    |  ✅  | ❌  |
| [getBaseOs()](#getbaseOs)                                         | `Promise<string>`   |  ❌  |   ✅    |   ✅    |  ✅  | ✅  |
| [getBuildId()](#getbuildid)                                       | `Promise<string>`   |  ✅  |   ✅    |   ✅    |  ✅  | ❌  |
| [getBatteryLevel()](#getbatterylevel)                             | `Promise<number>`   |  ✅  |   ✅    |   ✅    |  ✅  | ✅  |
| [getBootloader()](#getbootloader)                                 | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getBrand()](#getbrand)                                           | `string`            |  ✅  |   ✅    |   ✅    |  ✅  | ❌  |
| [getBuildNumber()](#getbuildnumber)                               | `string`            |  ✅  |   ✅    |   ✅    |  ✅  | ❌  |
| [getBundleId()](#getbundleid)                                     | `string`            |  ✅  |   ✅    |   ✅    |  ✅  | ❌  |
| [isCameraPresent()](#iscamerapresent)                             | `Promise<boolean>`  |  ❌  |   ✅    |   ✅    |  ✅  | ✅  |
| [getCarrier()](#getcarrier)                                       | `Promise<string>`   |  ✅  |   ✅    |   ❌    |  ❌  | ❌  |
| [getCodename()](#getcodename)                                     | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getDevice()](#getdevice)                                         | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getDeviceId()](#getdeviceid)                                     | `string`            |  ✅  |   ✅    |   ✅    |  ✅  | ❌  |
| [getDeviceType()](#getDeviceType)                                 | `string`            |  ✅  |   ✅    |   ❌    |  ✅  | ❌  |
| [getDisplay()](#getdisplay)                                       | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getDeviceName()](#getdevicename)                                 | `Promise<string>`   |  ✅  |   ✅    |   ✅    |  ✅  | ❌  |
| [getDeviceToken()](#getdevicetoken)                               | `Promise<string>`   |  ✅  |   ❌    |   ❌    |  ✅  | ❌  |
| [getFirstInstallTime()](#getfirstinstalltime)                     | `Promise<number>`   |  ✅  |   ✅    |   ✅    |  ❓  | ❌  |
| [getFingerprint()](#getfingerprint)                               | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❓  | ❌  |
| [getFontScale()](#getfontscale)                                   | `Promise<number>`   |  ✅  |   ✅    |   ✅    |  ❓  | ❌  |
| [getFreeDiskStorage()](#getfreediskstorage)                       | `Promise<number>`   |  ✅  |   ✅    |   ✅    |  ✅  | ✅  |
| [getFreeDiskStorageOld()](#getfreediskstorageold)                 | `Promise<number>`   |  ✅  |   ✅    |   ✅    |  ✅  | ✅  |
| [getHardware()](#gethardware)                                     | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getHost()](#gethost)                                             | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getIpAddress()](#getipaddress)                                   | `Promise<string>`   |  ✅  |   ✅    |   ✅    |  ✅  | ❌  |
| [getIncremental()](#getincremental)                               | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getInstallerPackageName()](#getinstallerpackagename)             | `Promise<string>`   |  ✅  |   ✅    |   ✅    |  ❓  | ❌  |
| [getInstallReferrer()](#getinstallreferrer)                       | `Promise<string>`   |  ❌  |   ✅    |   ✅    |  ❓  | ✅  |
| [getInstanceId()](#getinstanceid)                                 | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getLastUpdateTime()](#getlastupdatetime)                         | `Promise<number>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getMacAddress()](#getmacaddress)                                 | `Promise<string>`   |  ✅  |   ✅    |   ❌    |  ❓  | ❌  |
| [getManufacturer()](#getmanufacturer)                             | `Promise<string>`   |  ✅  |   ✅    |   ✅    |  ✅  | ❌  |
| [getMaxMemory()](#getmaxmemory)                                   | `Promise<number>`   |  ❌  |   ✅    |   ✅    |  ❓  | ✅  |
| [getModel()](#getmodel)                                           | `string`            |  ✅  |   ✅    |   ✅    |  ✅  | ❌  |
| [getPhoneNumber()](#getphonenumber)                               | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getPowerState()](#getpowerstate)                                 | `Promise<object>`   |  ✅  |   ✅    |   ✅    |  ❓  | ✅  |
| [getProduct()](#getproduct)                                       | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getPreviewSdkInt()](#getPreviewSdkInt)                           | `Promise<number>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getReadableVersion()](#getreadableversion)                       | `string`            |  ✅  |   ✅    |   ✅    |  ❓  | ❌  |
| [getSerialNumber()](#getserialnumber)                             | `Promise<string>`   |  ❌  |   ✅    |   ✅    |  ❓  | ❌  |
| [getSecurityPatch()](#getsecuritypatch)                           | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getSystemAvailableFeatures()](#getSystemAvailableFeatures)       | `Promise<string[]>` |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getSystemName()](#getsystemname)                                 | `string`            |  ✅  |   ✅    |   ✅    |  ❓  | ❌  |
| [getSystemVersion()](#getsystemversion)                           | `string`            |  ✅  |   ✅    |   ✅    |  ❓  | ❌  |
| [getTags()](#gettags)                                             | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getType()](#gettype)                                             | `Promise<string>`   |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [getTotalDiskCapacity()](#gettotaldiskcapacity)                   | `Promise<number>`   |  ✅  |   ✅    |   ✅    |  ❓  | ✅  |
| [getTotalDiskCapacityOld()](#gettotaldiskcapacityold)             | `Promise<number>`   |  ✅  |   ✅    |   ✅    |  ❓  | ✅  |
| [getTotalMemory()](#gettotalmemory)                               | `Promise<number>`   |  ✅  |   ✅    |   ❌    |  ❓  | ✅  |
| [getUniqueId()](#getuniqueid)                                     | `Promise<string>`   |  ✅  |   ✅    |   ✅    |  ✅  | ❌  |
| [getUsedMemory()](#getusedmemory)                                 | `Promise<number>`   |  ✅  |   ✅    |   ✅    |  ❓  | ✅  |
| [getUserAgent()](#getuseragent)                                   | `Promise<string>`   |  ✅  |   ✅    |   ❌    |  ❓  | ✅  |
| [getUserAgentSync()](#getuseragent)                               | `string`            |  ❌  |   ✅    |   ❌    |  ❌  | ✅  |
| [getVersion()](#getversion)                                       | `string`            |  ✅  |   ✅    |   ✅    |  ❓  | ❌  |
| [getBrightness()](#getBrightness)                                 | `Promise<number>`   |  ✅  |   ❌    |   ❌    |  ❓  | ❌  |
| [hasGms()](#hasGms)                                               | `Promise<boolean>`  |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [hasHms()](#hasHms)                                               | `Promise<boolean>`  |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [hasNotch()](#hasNotch)                                           | `boolean`           |  ✅  |   ✅    |   ✅    |  ❓  | ❌  |
| [hasDynamicIsland()](#hasDynamicIsland)                           | `boolean`           |  ✅  |   ✅    |   ✅    |  ❓  | ❌  |
| [hasSystemFeature()](#hassystemfeaturefeature)                    | `Promise<boolean>`  |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [isAirplaneMode()](#isairplanemode)                               | `Promise<boolean>`  |  ❌  |   ✅    |   ❌    |  ❌  | ✅  |
| [isBatteryCharging()](#isbatterycharging)                         | `Promise<boolean>`  |  ✅  |   ✅    |   ✅    |  ❓  | ✅  |
| [isEmulator()](#isemulator)                                       | `Promise<boolean>`  |  ✅  |   ✅    |   ✅    |  ❓  | ❌  |
| [isKeyboardConnected()](#iskeyboardconnected)                     | `Promise<bool>`     |  ❌  |   ❌    |   ✅    |  ❓  | ❌  |
| [isLandscape()](#isLandscape)                                     | `Promise<boolean>`  |  ✅  |   ✅    |   ✅    |  ❓  | ❌  |
| [isLocationEnabled()](#isLocationEnabled)                         | `Promise<boolean>`  |  ✅  |   ✅    |   ❌    |  ❓  | ✅  |
| [isMouseConnected()](#ismouseconneted)                            | `Promise<bool>`     |  ❌  |   ❌    |   ✅    |  ❓  | ❌  |
| [isHeadphonesConnected()](#isHeadphonesConnected)                 | `Promise<boolean>`  |  ✅  |   ✅    |   ❌    |  ❓  | ❌  |
| [isPinOrFingerprintSet()](#ispinorfingerprintset)                 | `Promise<boolean>`  |  ✅  |   ✅    |   ✅    |  ❓  | ❌  |
| [isTablet()](#istablet)                                           | `boolean`           |  ✅  |   ✅    |   ✅    |  ✅  | ❌  |
| [isTabletMode()](#istabletmode)                                   | `Promise<bool>`     |  ❌  |   ❌    |   ✅    |  ❌  | ❌  |
| [supported32BitAbis()](#supported32BitAbis)                       | `Promise<string[]>` |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [supported64BitAbis()](#supported64BitAbis)                       | `Promise<string[]>` |  ❌  |   ✅    |   ❌    |  ❌  | ❌  |
| [supportedAbis()](#supportedAbis)                                 | `Promise<string[]>` |  ✅  |   ✅    |   ✅    |  ✅  | ❌  |
| [syncUniqueId()](#syncuniqueid)                                   | `Promise<string>`   |  ✅  |   ❌    |   ❌    |  ✅  | ❌  |

---

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ❌     |
| Android |    ❌     |
| Windows |    ❌     |
| Macos  |    ✅     |
## Checklist

* [x] I have tested this on a device/simulator for each compatible OS
* [x] I added the documentation in `README.md`
* [x] I updated the typings files (`privateTypes.ts`, `types.ts`)
* [x] I added a sample use of the API (`example/App.js`)

### CODE REVIEW

**Comment by mikehardy** [comment]
I am closing and reopening this with only one goal: It should run the CI checks - I haven't reviewed but I want to see if the existing CI checks work on it

**Comment by fi-nik** [comment]
> I am closing and reopening this with only one goal: It should run the CI checks - I haven't reviewed but I want to see if the existing CI checks work on it

@mikehardy Ill fix these conflicts and check those failing checks

File: example/metro.config.js [hunk 1]
```diff
@@ -12,8 +12,19 @@ const rnwPath = fs.realpathSync(
   path.resolve(require.resolve('react-native-windows/package.json'), '..'),
 );
 
+const extraNodeModules = {
```
Comment (fi-nik) [inline example/metro.config.js:15]: Should be removed, leftover from testing phase

File: example/macos/example-macOS/example.entitlements [hunk 2]
```diff
@@ -2,7 +2,11 @@
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
-	<key>IDEDidComputeMac32BitWarning</key>
+	<key>com.apple.security.app-sandbox</key>
```
Comment (fi-nik) [inline example/macos/example-macOS/example.entitlements:5]: Is this really required?

**Comment by fi-nik** [comment]
@mikehardy I think i fixed issues with failing checks. Can someone rerun checks?

File: example/macos/example-macOS/example.entitlements (hunk 2 above)
Comment (mikehardy) [reply example/macos/example-macOS/example.entitlements:5]: for the example app....perhaps? I'm not an expert here but examples / dev builds usually go against apple's various sandbox systems

Comment (fi-nik) [reply example/macos/example-macOS/example.entitlements:5]: I wrote this comment more for myself, just to check is it required, should i leave it or remove it. It is leftover from previous work. 😁

**Comment by finik-htec** [comment]
@mikehardy Is there something blocking review of this PR ?

**Comment by cmgustavo** [comment]
Hi! Is there something that's blocking to merge this? 🙏🏼

**Comment by fi-nik** [comment]
@mikehardy @machour @GantMan @Johan-dutoit @schie This PR is opened too long. It's bit hard keeping it up to date with newest changes. Is there way for someone to review it, or to approve those tests? I could probably continue adding support for additional methods, but i think this is enough for start?

**Comment by mikehardy** [comment]
I keep dreaming I'll have time and I'm always sorry I haven't had time yet - the work has value though, undeniably

**Comment by parsable-FNikolic** [comment]
@mikehardy @machour  @GantMan  @Johan-dutoit @schie I can update PR for this again, but just wanna check with you if its going to merge any time soon? Can anyone do any review on this code, at least to know what else needed to be added? Its taking to long

**Comment by fi-nik** [comment]
@mikehardy  @machour  @GantMan  @Johan-dutoit @schie I can update PR for this again, but just wanna check with you if its going to merge any time soon? Can anyone do any review on this code, at least to know what else needed to be added?

**Comment by fi-nik** [comment]
Ah, im not sure what else i can do here :(

**Comment by schie** [comment]
@fi-nik , I can review this weekend, or early next week.

**Comment by fi-nik** [comment]
> @fi-nik , I can review this weekend, or early next week.

@schie Sure, its been long time a worked on this, i guess this is just part of work. Once this is merged i will check if there is more properties we can provide

**Pull request closed without merging**